lib/vrl/stdlib/src/map.rs

```diff
@@ -0,0 +1,164 @@
+        &[
+            Example {
+                title: "adding numbers in array",
+                source: r#"map([1,2,3]) -> |index, value| {  value + 1 }"#,
+                result: Ok("[2,3,4]"),
+            },
+            Example {
+                title: "enumerating object",
+                source: r#"map({"a" : 1, "b" : 2, "c" : 3}) -> |key, value| { [key, value + 1] }"#,
+                result: Ok(r#"{"a" :2, "b" : 3, "c" :4}"#),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::Compiler,
+        _ctx: &vrl::prelude::FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(MapFn { value, closure }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        let object = closure::Input {
+            parameter_keyword: "value",
+            kind: Kind::Object,
+            variables: vec![
+                closure::Variable { kind: Kind::Bytes },
+                closure::Variable { kind: Kind::all() },
+            ],
+            output: closure::Output::Array {
+                elements: vec![Kind::Bytes, Kind::all()],
+            },
+        };
+
+        let array = closure::Input {
+            parameter_keyword: "value",
+            kind: Kind::Array,
+            variables: vec![
+                closure::Variable {
+                    kind: Kind::Integer,
+                },
+                closure::Variable { kind: Kind::all() },
+            ],
+            output: closure::Output::Any,
+        };
+
+        Some(closure::Definition {
+            inputs: vec![object, array],
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct MapFn {
+    value: Box<dyn Expression>,
+    closure: Closure,
+}
+
+impl Expression for MapFn {
+    fn resolve(&self, ctx: &mut Context) -> Result<Value> {
+        // - First focus on objects, ignore arrays
+        // - A closure can resolve if it knows:
+        //   - Closure variable signature/identifiers
+        //   - Content of the object
+        //   - Access to `ctx`
+        //
+        // - Resolving closure should be agnostic. It takes the values you want to assign to
+        //   different closure variables, and it takes an `Fn` to apply to the data.
+
+        let mut object_result: BTreeMap<String, Value> = BTreeMap::default();
+        let mut array_result: Vec<Value> = Vec::default();
+        let mut is_array = false;
+        let value = self.value.resolve(ctx)?;
+
+        let mut map = |_: &Context, output: Output| -> Result<()> {
+            match output {
+                Output::Object { key, value } => {
+                    object_result.insert(key, value);
+                }
+                Output::Array { element } => {
+                    array_result.push(element);
+                    is_array = true;
+                }
+            };
```

Comment:
I realize this is totally hokey, it's just a workaround until I can change the closure's `resolve` method.